// @dbml/core/lib/export/MysqlExporter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var MySQLExporter =
/*#__PURE__*/
function () {
  function MySQLExporter() {
    _classCallCheck(this, MySQLExporter);
  }

  _createClass(MySQLExporter, null, [{
    key: "getFieldLines",
    value: function getFieldLines(tableId, model) {
      var table = model.tables[tableId];
      var lines = table.fieldIds.map(function (fieldId) {
        var field = model.fields[fieldId];
        var line = '';

        if (field.enumId) {
          var _enum = model.enums[field.enumId];
          line = "`".concat(field.name, "` ENUM (");

          var enumValues = _enum.valueIds.map(function (valueId) {
            return "'".concat(model.enumValues[valueId].name, "'");
          });

          line += "".concat(enumValues.join(', '), ")");
        } else {
          line = "`".concat(field.name, "` ").concat(field.type.type_name !== 'varchar' ? field.type.type_name : 'varchar(255)');
        }

        if (field.unique) {
          line += ' UNIQUE';
        }

        if (field.pk) {
          line += ' PRIMARY KEY';
        }

        if (field.not_null) {
          line += ' NOT NULL';
        }

        if (field.increment) {
          line += ' AUTO_INCREMENT';
        }

        if (field.dbdefault) {
          if (field.dbdefault.type === 'expression') {
            line += " DEFAULT (".concat(field.dbdefault.value, ")");
          } else if (field.dbdefault.type === 'string') {
            line += " DEFAULT \"".concat(field.dbdefault.value, "\"");
          } else {
            line += " DEFAULT ".concat(field.dbdefault.value);
          }
        }

        if (field.note) {
          line += " COMMENT '".concat(field.note.replace(/'/g, "\\'"), "'");
        }

        return line;
      });
      return lines;
    }
  }, {
    key: "getCompositePKs",
    value: function getCompositePKs(tableId, model) {
      var table = model.tables[tableId];
      var compositePkIds = table.indexIds ? table.indexIds.filter(function (indexId) {
        return model.indexes[indexId].pk;
      }) : [];
      var lines = compositePkIds.map(function (keyId) {
        var key = model.indexes[keyId];
        var line = 'PRIMARY KEY';
        var columnArr = [];
        key.columnIds.forEach(function (columnId) {
          var column = model.indexColumns[columnId];
          var columnStr = '';

          if (column.type === 'expression') {
            columnStr = "(".concat(column.value, ")");
          } else {
            columnStr = "`".concat(column.value, "`");
          }

          columnArr.push(columnStr);
        });
        line += " (".concat(columnArr.join(', '), ")");
        return line;
      });
      return lines;
    }
  }, {
    key: "getTableContentArr",
    value: function getTableContentArr(tableIds, model) {
      var tableContentArr = tableIds.map(function (tableId) {
        var fieldContents = MySQLExporter.getFieldLines(tableId, model);
        var compositePKs = MySQLExporter.getCompositePKs(tableId, model);
        return {
          tableId: tableId,
          fieldContents: fieldContents,
          compositePKs: compositePKs
        };
      });
      return tableContentArr;
    }
  }, {
    key: "exportTables",
    value: function exportTables(tableIds, model) {
      var tableContentArr = MySQLExporter.getTableContentArr(tableIds, model);
      var tableStrs = tableContentArr.map(function (tableContent) {
        var content = [].concat(_toConsumableArray(tableContent.fieldContents), _toConsumableArray(tableContent.compositePKs));
        var table = model.tables[tableContent.tableId];
        var schema = model.schemas[table.schemaId];
        var tableStr = "CREATE TABLE ".concat((0, _utils.shouldPrintSchema)(schema, model) ? "`".concat(schema.name, "`.") : '', "`").concat(table.name, "` (\n").concat(content.map(function (line) {
          return "  ".concat(line);
        }).join(',\n'), "\n);\n");
        return tableStr;
      });
      return tableStrs.length ? tableStrs.join('\n') : '';
    }
  }, {
    key: "buildFieldName",
    value: function buildFieldName(fieldIds, model) {
      var fieldNames = fieldIds.map(function (fieldId) {
        return "`".concat(model.fields[fieldId].name, "`");
      }).join(', ');
      return "(".concat(fieldNames, ")");
    }
  }, {
    key: "exportRefs",
    value: function exportRefs(refIds, model) {
      var _this = this;

      var strArr = refIds.map(function (refId) {
        var ref = model.refs[refId];
        var refEndpointIndex = ref.endpointIds.findIndex(function (endpointId) {
          return model.endpoints[endpointId].relation === '1';
        });
        var foreignEndpointId = ref.endpointIds[1 - refEndpointIndex];
        var refEndpointId = ref.endpointIds[refEndpointIndex];
        var foreignEndpoint = model.endpoints[foreignEndpointId];
        var refEndpoint = model.endpoints[refEndpointId];
        var refEndpointField = model.fields[refEndpoint.fieldIds[0]];
        var refEndpointTable = model.tables[refEndpointField.tableId];
        var refEndpointSchema = model.schemas[refEndpointTable.schemaId];

        var refEndpointFieldName = _this.buildFieldName(refEndpoint.fieldIds, model, 'mysql');

        var foreignEndpointField = model.fields[foreignEndpoint.fieldIds[0]];
        var foreignEndpointTable = model.tables[foreignEndpointField.tableId];
        var foreignEndpointSchema = model.schemas[foreignEndpointTable.schemaId];

        var foreignEndpointFieldName = _this.buildFieldName(foreignEndpoint.fieldIds, model, 'mysql');

        var line = "ALTER TABLE ".concat((0, _utils.shouldPrintSchema)(foreignEndpointSchema, model) ? "`".concat(foreignEndpointSchema.name, "`.") : '', "`").concat(foreignEndpointTable.name, "` ADD ");

        if (ref.name) {
          line += "CONSTRAINT `".concat(ref.name, "` ");
        }

        line += "FOREIGN KEY ".concat(foreignEndpointFieldName, " REFERENCES ").concat((0, _utils.shouldPrintSchema)(refEndpointSchema, model) ? "`".concat(refEndpointSchema.name, "`.") : '', "`").concat(refEndpointTable.name, "` ").concat(refEndpointFieldName);

        if (ref.onDelete) {
          line += " ON DELETE ".concat(ref.onDelete.toUpperCase());
        }

        if (ref.onUpdate) {
          line += " ON UPDATE ".concat(ref.onUpdate.toUpperCase());
        }

        line += ';\n';
        return line;
      });
      return strArr.length ? strArr.join('\n') : '';
    }
  }, {
    key: "exportIndexes",
    value: function exportIndexes(indexIds, model) {
      // exclude composite pk index
      var indexArr = indexIds.filter(function (indexId) {
        return !model.indexes[indexId].pk;
      }).map(function (indexId, i) {
        var index = model.indexes[indexId];
        var table = model.tables[index.tableId];
        var schema = model.schemas[table.schemaId];
        var line = 'CREATE';

        if (index.unique) {
          line += ' UNIQUE';
        }

        var indexName = index.name ? "`".concat(index.name, "`") : "`".concat((0, _utils.shouldPrintSchema)(schema, model) ? "`".concat(schema.name, "`.") : '').concat(table.name, "_index_").concat(i, "`");
        line += " INDEX ".concat(indexName, " ON ").concat((0, _utils.shouldPrintSchema)(schema, model) ? "`".concat(schema.name, "`.") : '', "`").concat(table.name, "`");
        var columnArr = [];
        index.columnIds.forEach(function (columnId) {
          var column = model.indexColumns[columnId];
          var columnStr = '';

          if (column.type === 'expression') {
            columnStr = "(".concat(column.value, ")");
          } else {
            columnStr = "`".concat(column.value, "`");
          }

          columnArr.push(columnStr);
        });
        line += " (".concat(columnArr.join(', '), ")");

        if (index.type) {
          line += " USING ".concat(index.type.toUpperCase());
        }

        line += ';\n';
        return line;
      });
      return indexArr.length ? indexArr.join('\n') : '';
    }
  }, {
    key: "exportComments",
    value: function exportComments(comments, model) {
      var commentArr = comments.map(function (comment) {
        var line = '';

        if (comment.type === 'table') {
          var table = model.tables[comment.tableId];
          line += "ALTER TABLE `".concat(table.name, "` COMMENT = '").concat(table.note.replace(/'/g, "\\'"), "'");
        }

        line += ';\n';
        return line;
      });
      return commentArr.length ? commentArr.join('\n') : '';
    }
  }, {
    key: "export",
    value: function _export(model) {
      var res = '';
      var hasBlockAbove = false;
      var database = model.database['1'];
      var indexIds = [];
      var comments = [];
      database.schemaIds.forEach(function (schemaId) {
        var schema = model.schemas[schemaId];
        var tableIds = schema.tableIds,
            refIds = schema.refIds;

        if ((0, _utils.shouldPrintSchema)(schema, model)) {
          if (hasBlockAbove) res += '\n';
          res += "CREATE DATABASE `".concat(schema.name, "`;\n");
          hasBlockAbove = true;
        }

        if (!_lodash["default"].isEmpty(tableIds)) {
          if (hasBlockAbove) res += '\n';
          res += MySQLExporter.exportTables(tableIds, model);
          hasBlockAbove = true;
        }

        if (!_lodash["default"].isEmpty(refIds)) {
          if (hasBlockAbove) res += '\n';
          res += MySQLExporter.exportRefs(refIds, model);
          hasBlockAbove = true;
        }

        indexIds.push.apply(indexIds, _toConsumableArray(_lodash["default"].flatten(tableIds.map(function (tableId) {
          return model.tables[tableId].indexIds;
        }))));
        comments.push.apply(comments, _toConsumableArray(_lodash["default"].flatten(tableIds.map(function (tableId) {
          var note = model.tables[tableId].note;
          return note ? [{
            type: 'table',
            tableId: tableId
          }] : [];
        }))));
      });

      if (!_lodash["default"].isEmpty(indexIds)) {
        if (hasBlockAbove) res += '\n';
        res += MySQLExporter.exportIndexes(indexIds, model);
        hasBlockAbove = true;
      }

      if (!_lodash["default"].isEmpty(comments)) {
        if (hasBlockAbove) res += '\n';
        res += MySQLExporter.exportComments(comments, model);
        hasBlockAbove = true;
      }

      return res;
    }
  }]);

  return MySQLExporter;
}();

var _default = MySQLExporter;
exports["default"] = _default;
