// @dbml/core/lib/parse/mssql/utils.js
"use strict";

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var P = require('parsimmon');

var _require = require('./keyword_parsers'),
    LParen = _require.LParen,
    RParen = _require.RParen,
    Comma = _require.Comma;

var wss = require('./whitespaces');

exports.makeNode = function () {
  return function (parser) {
    return P.seqMap(P.index, parser, P.index, function (start, value, end) {
      if (!value) return parser;
      if (_typeof(value.value) !== 'object') value.value = {
        value: value.value
      };
      value.value.token = {
        start: start,
        end: end
      };
      return value;
    }).skip(wss);
  };
};

exports.makeList = function (parser) {
  var isZero = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  var seperator = parser.sepBy1(Comma);
  if (isZero) seperator = parser.sepBy(Comma);
  return P.seq(LParen, seperator, RParen).map(function (value) {
    return value[1];
  });
};

exports.streamline = function (type) {
  return function (parser) {
    return parser.skip(wss).map(function (value) {
      // eslint-disable-next-line no-param-reassign
      if (value !== 0 && !value) value = '';
      return {
        type: type,
        value: value
      };
    });
  };
};
